arangod/Cluster/ClusterInfo.cpp

```diff
@@ -513,6 +515,124 @@ void ClusterInfo::triggerBackgroundGetIds() {
   }
 }
 
+auto ClusterInfo::createReplicatedStateSpec(
+    std::string const& shardId, std::vector<std::string> const& serverIds,
+    std::size_t writeConcern, std::size_t softWriteConcern)
+    -> replication2::replicated_state::agency::Target {
+  replication2::replicated_state::agency::Target spec;
+
+  spec.id = LogicalCollection::shardIdToLogId(shardId);
+  spec.properties.implementation.type = "black-hole";
+  TRI_ASSERT(!serverIds.empty());
+  spec.leader = serverIds.front();
+
+  for (auto const& serverId : serverIds) {
+    spec.participants.emplace(
+        serverId,
+        replication2::replicated_state::agency::Target::Participant{});
+  }
+
+  spec.config.writeConcern = writeConcern;
+  spec.config.softWriteConcern = softWriteConcern;
+  spec.config.waitForSync = false;
+  spec.version = 1;
+
+  return spec;
+}
+
+auto ClusterInfo::waitForReplicatedStatesCreation(
+    std::string const& databaseName,
+    std::vector<replication2::replicated_state::agency::Target> const&
+        replicatedStates) -> futures::Future<Result> {
+  auto replicatedStateMethods =
+      arangodb::replication2::ReplicatedStateMethods::createInstanceCoordinator(
+          _server, databaseName);
+
+  std::vector<futures::Future<ResultT<consensus::index_t>>> futureStates;
+  futureStates.reserve(replicatedStates.size());
+  for (auto const& spec : replicatedStates) {
+    futureStates.emplace_back(
+        replicatedStateMethods->waitForStateReady(spec.id, *spec.version));
+  }
+
+  return futures::collectAll(std::move(futureStates))
+      .thenValue(
+          [&clusterInfo = _server.getFeature<ClusterFeature>().clusterInfo()](
+              auto&& raftIndices) {
+            consensus::index_t highestRaftIndex = std::transform_reduce(
+                raftIndices.begin(), raftIndices.end(), (consensus::index_t)0,
+                [](consensus::index_t l, consensus::index_t r) {
+                  return std::max(l, r);
+                },
+                [](auto&& value) { return value.get().get(); });
```

Comment:
I know I was the one who wrote this `transform_reduce`, but maybe a simple loop would be easier to read.
```suggestion
            consensus::index_t maxIndex = 0;
            for (auto& v : raftIndices) {
              maxIndex = std::max(maxIndex, v.get().get());
            }
```
Also, what if one of the `raftIndices` entries contains an error?

Reply:
Then we're going to handle the error inside the `.then()`, but that would cause the function to return a `Result` containing a failure. We could choose to ignore a failed `raftIndex`, but it is a delicate situation, because we might discard the maximum.